### PR TITLE
rewriter: Rewrote the meson rewriter - now works with AST modification

### PR DESCRIPTION
**Pull request by mensinda**
This PR replaces the old rewriter and moves some common functionality into a separate ast package.

The new rewriter works by modifying the AST instead of modifying the meson files directly. Also, only the modified parts of the AST are overwritten. This helps to preserve comments and indentation.

The rewriter currently only supports adding and removing source files from a target. However, this PR also provides the framework to easily modify other functions, such as `project` and `dependency`.

The rewriter is still in beta, so I didn't add a release snippet.

cc @textshell

### CODE REVIEW

File: mesonbuild/rewriter.py [hunk 1]
```diff
@@ -23,36 +23,301 @@
...
+            elif isinstance(n, mparser.ArrayNode):
+                args = n.args.arguments
+            elif isinstance(n, mparser.ArgumentNode):
+                args = n.arguments
+            return args
+
+        if cmd['operation'] == 'src_add':
+            #################################
+            ### Add sources to the target ###
+            #################################
```
Comment (jpakkane) [inline mesonbuild/rewriter.py]: Don't do this. There is no reason to have shouty comment syntax.

File: mesonbuild/ast/printer.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            level = node.level
+        else:
+            level = self.last_level
+        self.last_level = level
+        if self.is_newline:
+            self.result += ' ' * (level * self.indent)
+        self.result += data
+        self.is_newline = False
+
+    def appendS(self, data: str, node: mparser.BaseNode):
```
Comment (jpakkane) [inline mesonbuild/ast/printer.py]: This method could have a clearer name. It's not obvious what `appendS` stands for.

**Comment by jpakkane** [comment]
Overall, good stuff. Great that this part is being brought back to life.

File: mesonbuild/rewriter.py (hunk 1 above)
Comment (mensinda) [reply mesonbuild/rewriter.py]: Sure, removed with the last commit.

File: mesonbuild/ast/printer.py (hunk 2 above)
Comment (mensinda) [reply mesonbuild/ast/printer.py]: Renamed the function to `append_padded`.

**Pull request merged**